### ISSUE TITLE
OneDive Plugin

### ISSUE THREAD

**Issue by orangeberry** [syncany/syncany#454]
Support for Microsoft OneDrive would be nice.
I will try to implement such a plugin myself, but I don't know if I can do that

**Comment by cr0** [comment]
see https://github.com/syncany/syncany/wiki/Plugin-Overview

**Comment by J0s3f** [comment]
Just for reference: The Documentation is here https://dev.onedrive.com/README.htm

One project with good OneDrive support is the backup software duplicati https://github.com/duplicati/duplicati/tree/master/Duplicati/Library/Backend/OneDrive

**Comment by dewdad** [comment]
+1

**Comment by Nitrogenycs** [comment]
I was pondering to implement this plugin and then came across the requirement to register your app with microsoft first. Wondering how the google plugin did this I took a look at it and ended up at https://github.com/Lyrrad/syncany-plugin-googledrive/blob/master/src/main/java/org/syncany/plugins/googledrive/GoogledriveTransferPlugin.java .
Is it really a good idea to store the app id and secret in there? Does syncany have a procedure to deal with this kind of problem of having to register an app id?

**Comment by Nitrogenycs** [comment]
@melect: any progress with your implementation so far you want to share?

**Comment by pimotte** [comment]
Yeah. I'm not sure why we did that in the first place. There's the possibility of using encrypted environment variables in travis: https://docs.travis-ci.com/user/environment-variables/ We use these for the API key to upload binaries: https://travis-ci.org/syncany/syncany/builds/105397826.

The thing is that using that helps a little bit (namely against people searching github for keys), but it will still be trivial to decompile the jar and extract the keys that way. Which is the reason why we didn't do it in the first place, but it's probably worth the effort to use the enviroment variables for new ones.

**Comment by melect** [comment]
@Nitrogenycs: Unfortunately the most of my implementation got lost, due a hard drive error. 
My biggest problem was the handling with the authentication process, because I din't know a way to get a refresh token without the use of a browser. 
My solution then was to write an external program which generates a link to get the refresh token. This was't really suitable.
The other problem (as you mentioned) was the problem with the app id/secret. I don't found a way to store keys secure in a jar.

If there is a good solution for the authentication and for the app id/secret, the other parts won't be to difficulty. 
In my testing the upload and download through the OneDrive API wasn't a problem.

**Comment by franklington** [comment]
+1
